src/validation.cpp

```diff
@@ -3440,8 +3440,22 @@ void Chainstate::ResetBlockFailureFlags(CBlockIndex *pindex) {
     }
 }
 
+void Chainstate::TryAddBlockIndexCandidate(CBlockIndex* pindex)
+{
+    // FIXME: the background-validation chainstate should only add new entries
+    // that build on blocks for which we actually have all the data, while the
+    // snapshot chainstate (if we have one) is able to build on blocks that are
+    // missing but for which BLOCK_ASSUME_VALID is set.
+    // So this is broken right now. XXX
```

Comment:
Sorry that comment was confusing.  I dropped it, and instead change the behavior around managing `setBlockIndexCandidates` in the last two commits. 

I think master has a couple bugs. When we start up, in `LoadBlockIndex()` we weren't adding blocks that have been downloaded and have more work than the background chainstate's tip to `setBlockIndexCandidates`, which prevents those blocks from being validated and connected on startup.  Also, because `AcceptBlock()` is a member of `Chainstate`, it's at best confusing to reason about whether a block that would advance the tip of a chainstate would always be added to a chainstate's `setBlockIndexCandidates` (it depends on yet-more logic around where a block came from and guessing about which chainstate might need it).

The new code tries to simplify things by establishing that the background chainstate is only trying to connect blocks towards the snapshot block, and that all blocks for which we HAVE_DATA and have more work than the background chainstate's tip (and which are ancestors of the snapshot block) will be added to the background chainstate's `setBlockIndexCandidates`.

Reply:
I guess the fairest thing to say is that the changes to `AcceptBlock()` are to prevent a future bug -- it's possible that we could write code that somehow gets it right, but the design of selectively adding blocks to one chainstate's block index, and not the other, is one that I found very confusing. 

As an example, if you took #24008 (which proposed logic for determining which chainstate to invoke AcceptBlock() on) and combined that with changing the logic in `AcceptBlock()` so that the background chainstate wouldn't bother processing blocks that are past the snapshot base, then that would be problematic if we were to allow pruning the snapshot-based-chain while the background sync is running (because according to the logic in #24008, if the block is already on the active chain, then we give it to the background chainstate to process).  That's 2 hypotheticals so certainly this is speculative, but I think that those changes would all make sense on their own, so the fact that they fail to compose with the code in master is a sign to me that we should take a different design.